### PR TITLE
Implements opt-in behavior

### PR DESCRIPTION
**Pull request by bravo-kernel**
This PR solves the issues described in #7 by moving from opt-out to opt-in using two custom finders. 

**Two custom finders:**
I had to implement a second custom finder (and rename the existing one) to cater for all scenario's:
- `findObfuscated`: use to find records using an obfuscated primary key
- `findObfuscate`: use to obfuscate primary keys in returned result set

**Default finds**
Solves the issues described in #7 because without the opt-in nothing extra happens. https://github.com/UseMuffin/Obfuscate/pull/8/files#diff-aad0cd1f077b7222a4d9886d1a9a7e49R92

**Auth**
Solves (JWT) authentication by specifying the `obfuscated` finder in the authentication config options:

``` php
           'authenticate' => [
                'Form' => [
                    'fields' => ['username' => 'email', 'password' => 'password']
                ],
                'ADmad/JwtAuth.Jwt' => [
                    'finder' => 'obfuscated',
                    'userModel' => 'Users',
                    'fields' => [
                        'username' => 'id'
                    ],
                    'parameter' => 'token',
                    'queryDatasource' => true // must be true, see docblock
                ]
        ]
```

**Obfuscated index action**
This now works for me:

``` php
    public function index()
    {
        $this->Crud->on('beforePaginate', function (Event $event) {
            $event->subject()->query->find('obfuscate');
        });
        return $this->Crud->execute();
}
```

**Obfuscated view action**
This now also works for me:

``` php
    public function view()
    {
        $this->Crud->on('beforeFind', function (Event $event) {
            $event->subject()->query->find('obfuscate');
        });
        return $this->Crud->execute();
    }
```

Feedback welcome.

### CODE REVIEW

File: src/Model/Behavior/ObfuscateBehavior.php [hunk 1]
```diff
@@ -88,6 +89,10 @@ public function afterSave(Event $event, EntityInterface $entity, ArrayObject $op
      */
     public function beforeFind(Event $event, Query $query, ArrayObject $options, $primary)
     {
+        if (!isset($options['obfuscate'])) {
```
Comment (ADmad) [inline src/Model/Behavior/ObfuscateBehavior.php]: This should be `if (empty($options['obfuscate']))`.  Also you can combine the `!$primary` check.

**Comment by ADmad** [comment]
Shouldn't the `view()` method also expect IDs in url to be obfuscated and hence use `'obfuscated'` and `'obfucate'` both?

**Comment by bravo-kernel** [comment]
I thought that too but it turns out this "just works". If I pass an obfuscated id to the view action combined with the view() function I get both the correct record and the response contains the obfuscated id as well.

**Comment by ADmad** [comment]
Okay :)

**Comment by ADmad** [comment]
While both finder names `obfucate` and `obfucated` are grammatically correct for what they do, they are pretty similar and non-english speakers might be confused. But I don't have any ideas for better naming :)

**Comment by bravo-kernel** [comment]
I agree on the naming and actually spent some time on alternatives but they all ended up with the same problem; the presence/absence of the trailing `d` holds the meaning. All others were unusable (to me):

```
$this->find('applyObfuscate');
$this->find('ObfuscateKeys)
etc
```

After using it I now actually quite like the current finder names.

**Comment by jadb** [comment]
Very nice @bravo-kernel!

Ok, ok.. took me some time to wrap my head around why @ADmad suggested going this route (sorry hadn't replied back on the other thread) but I understand better now (at least I think). 

The naming didn't help either but like you I wasn't able to think of anything better; for this reason, I would like to have some documentation added before merging this and mixing up even more people.

What do you guys think?

**Comment by ADmad** [comment]
Yup some updates to readme would be nice.

**Comment by bravo-kernel** [comment]
Good to hear, I will update the readme with some basic instructions. The full rewrite will be a separate PR.

**Comment by bravo-kernel** [comment]
Updated docs with basic instructions. One possible naming alternative came to mind (cloak/cloaked) but I think we should stick to the proper and current names. The docs contain the cloak terms to clarify what they do for new users.

File: README.md [hunk 2]
```diff
@@ -5,7 +5,7 @@
 [![Total Downloads](https://img.shields.io/packagist/dt/muffin/obfuscate.svg?style=flat-square)](https://packagist.org/packages/muffin/obfuscate)
 [![License](https://img.shields.io/badge/license-MIT-blue.svg?style=flat-square)](LICENSE)
 
-CakePHP 3 support for ID obfuscation.
+CakePHP 3 support for ID obfuscation using HashIds, Optimus and Tiny.
```
Comment (jadb) [inline README.md]: Replace `and` by `or` and if we are to mention them, maybe mention also or any custom strategy.

**Comment by jadb** [comment]
:+1:

File: README.md (hunk 2 above)
Comment (bravo-kernel) [inline README.md]: Changed as requested. Additional info will be in the separate PR

**Comment by bravo-kernel** [comment]
:shipit:

**Comment by jadb** [comment]
You didn't add custom strategies, but I guess it can wait for the doc re-write ;)

**Comment by bravo-kernel** [comment]
One could say I have been trained well... clean separation, no PR creep :dancer:

**Comment by jadb** [comment]
:metal: :ok_hand:
